File: leetcode/C++/279.cpp
#include <vector>
#include <utility>
using namespace std;

class Solution
{
  public:
    vector<int> v;
    int numSquares(int n)
    {
        v.resize(n + 1);
        for (int i = 0; i <= n; ++i)
        {
            v[i] = INT_MAX;
        }
        for (int i = 1; i <= sqrt(n); ++i)
        {
            v[i * i] = 1;
        }
        for (int i = 1; i <= n; ++i)
        {
            for (int j = 1; i + j * j <= n; ++j)
            {
                v[i + j * j] = minInt(v[i] + 1, v[i + j * j]);
            }
        }
        return v[n];
    }
    int minInt(int a, int b)
    {
        return a < b ? a : b;
    }
};

File: leetcode/Go/279.go
package main

import "math"

var arr []int

func numSquares(n int) int {
	arr = make([]int, n+1)
	for i := 0; i < n+1; i++ {
		arr[i] = math.MaxInt32
	}
	for i := 1; float64(i) <= math.Sqrt(float64(n)); i++ {
		arr[i*i] = 1
	}

	for i := 1; i <= n; i++ {
		for j := 1; i+j*j <= n; j++ {
			arr[i+j*j] = minInt(arr[i]+1, arr[i+j*j])
		}
	}
	return arr[n]
}
